src/UnscentedTransformKalmanInversion.jl

```diff
@@ -0,0 +1,828 @@
+        else
+            tmp = ug_cov / gg_cov
+            u_mean = u_p_mean + tmp * (obs_mean - g_mean)
+            uu_cov = uu_p_cov - tmp * ug_cov'
+        end
+        =#
+
+        # sqrt-increments
+        X = FT.((u_p .- u_p_mean) / sqrt(m - 1))
+        Y = FT.((g_ext .- g_mean_ext) / sqrt(m - 1)) # may cause issue as first row = 0? TBD
+
+        # Create/Enlarge buffers if needed
+        tmp = get_buffer(get_process(uki)) # the buffer stores Y' * Γ_inv of [size(Y,2),size(Y,1)]
+        ys1, ys2 = size(Y)
+        if length(tmp) == 0  # no buffer
+            push!(tmp, zeros(ys2, ys1)) # stores Y' * Γ_inv
+            push!(tmp, zeros(ys2, ys2)) # stores Y' * Γ_inv * Y
+        elseif (size(tmp[1], 1) < ys2) || (size(tmp[1], 2) < ys1) # existing buffer is too small
+            tmp[1] = zeros(ys2, ys1)
+            tmp[2] = zeros(ys2, ys2)
+        end
+
+        lmul_obs_noise_cov_inv!(view(tmp[1]', :, 1:ys2), uki, Y, onci_idx) # store in transpose, with view helping reduce allocations
+        view(tmp[1], 1:ys2, :) .*= inv_noise_scaling
+
+        tmp[2][1:ys2, 1:ys2] = tmp[1][1:ys2, 1:ys1] * Y
+
+        for i in 1:ys2
+            tmp[2][i, i] += 1.0
+        end
+        Ω = inv(tmp[2][1:ys2, 1:ys2]) # Ω = (I + Y' * Γ_inv * Y)^-1 = I - Y' (Y Y' + Γ_inv)^-1 Y
+        u_mean = u_p_mean + X * FT.(Ω * tmp[1][1:ys2, 1:ys1] * (y_ext .- g_mean_ext)) #  mean update = Ω * Y' * Γ_inv * (y .- g_mean))
+        uu_cov = X * Ω * X' # cov update 
+
+
+        ########### Save results
+        push!(process.obs_pred, g_mean) # N_ens x N_data
+        push!(process.u_mean, u_mean) # N_ens x N_params
+        push!(process.uu_cov, uu_cov) # N_ens x N_data
+
+    end
+    function failsafe_update(uki, u, g, u_idx, g_idx, obs_noise_cov_inv, onci_idx, failed_ens, prior_mean)
+        #perform analysis on the model runs
+        succ_gauss_analysis!(uki, u, g, u_idx, g_idx, obs_noise_cov_inv, onci_idx, failed_ens, prior_mean)
+        #perform new prediction output to model parameters u_p
+        u_p = update_ensemble_prediction!(process, get_Δt(uki)[end], u_idx)
+        return u_p
+    end
+    return FailureHandler{TransformUnscented, SampleSuccGauss}(failsafe_update)
+end
+
+"""
+$(TYPEDSIGNATURES)
+
+Construct the sigma ensemble based on the mean `x_mean` and covariance `x_cov`.
+"""
+function construct_sigma_ensemble(
+    process::TransformUnscented,
+    x_mean::AbstractVector{FT},
+    x_cov::AbstractMatrix{FT},
+) where {FT <: AbstractFloat}
+
+    N_x = size(x_mean, 1)
+    N_ens = process.N_ens
+    c_weights = process.c_weights
+
+    # compute cholesky factor L of x_cov
+    local chol_xx_cov
+    try
+        chol_xx_cov = cholesky(Hermitian(x_cov)).L
+    catch
+        _, S, Ut = svd(x_cov)
+        # find the first singular value that is smaller than 1e-8
+        ind_0 = searchsortedfirst(S, 1e-8, rev = true)
+        S[ind_0:end] .= S[ind_0 - 1]
+        chol_xx_cov = (qr(sqrt.(S) .* Ut).R)'
```

Comment:
Do you need the QR factorization here? You only really need this to be the sqrt of the covariance matrix, right? So it doesn't need to be triangular, does it?

Reply:
As this is not something we have changed for the UTKI (it was copied from the implementation in UKI) . I will create a separate issue for it! I do agree though.